irohad/pending_txs_storage/impl/pending_txs_storage_impl.cpp

```diff
@@ -69,28 +69,34 @@ PendingTransactionStorageImpl::getPendingTransactions(
          and remaining_space >= batch_iterator->get()->transactions().size()) {
     auto &txs = batch_iterator->get()->transactions();
     // can be easily extended by adding more conditions if needed
-    if (first_tx_time or last_tx_time) {
-      auto time_predicate = [&](auto &tx) {
-        bool first_tx_time_valid =
-            first_tx_time ? first_tx_time < tx->createdTime() : true;
-        bool last_tx_time_valid =
-            last_tx_time ? tx->createdTime() < last_tx_time : true;
-        return first_tx_time_valid && last_tx_time_valid;
+    std::function<bool(int64_t)> time_pred_f;
+    if (first_tx_time) {
+      time_pred_f = [&first_tx_time](int64_t tx_tms) {
+        return tx_tms > first_tx_time;
       };
-      auto filtered_txs_size = std::count_if(txs.begin(), txs.end(), time_predicate);
-      std::copy_if(txs.begin(),
-                   txs.end(),
-                   std::back_inserter(response.transactions),
-                   time_predicate);
-      remaining_space -= filtered_txs_size;
-      ++batch_iterator;
     } else {
-      response.transactions.insert(
-          response.transactions.end(), txs.begin(), txs.end());
-      remaining_space -= txs.size();
-      ++batch_iterator;
+      time_pred_f = [](auto tx_tms) { return true; };
     }
-    
+    std::function<bool(int64_t)> time_pred_l;
+    if (last_tx_time) {
+      time_pred_l = [&last_tx_time](int64_t tx_tms) {
+        return tx_tms < last_tx_time;
+      };
+    } else {
+      time_pred_l = [](auto tx_tms) { return true; };
+    }
+    auto time_predicate = [&time_pred_f, &time_pred_l](auto &tx) {
+      auto created_time = tx->createdTime();
+      return time_pred_f(created_time) and time_pred_l(created_time);
+    };
```

Comment:
what about something like that:
```
auto time_predicate = [first_tx_time, last_tx_time](const auto& tx) {
     auto time_pred_f = first_tx_time ? tx_tms > first_tx_time : true;
     auto time_pred_l = last_tx_time ? tx_tms < last_tx_time : true;
     return time_pred_f && time_pred_l;
}
```
and You will not need lines [72, 87] and less if-else statements, shorted code and no `std::function`.
Don't worry about performance in the situation, because `:?` is in many compilers is optimized by calculating both patches at once, there is also predictor in CPU, which can predict that if `first_tx_time` was `false` few times it is more probably that it will be `false` again. (@kuvaldini if You don't agree with this please let me know, I still don't know much about performance)

Reply:
If we want to avoid checking those flags in all transactions we can try:
```
auto time_predicate_f = first_tx_time ? [first_tx_time](const auto& tx){ return first_tx_time < tx->createdTime(); } : [](const auto&){return true;};
auto time_predicate_l = last_tx_time ? [last_tx_time](const auto& tx){ return tx->createdTime() < last_tx_time); } : [](const auto&){return true;};
auto time_predicate = [time_predicate_f, time_predicate_l](tx){ return time_predicate_f(tx) && time_predicate_l(tx); }
```